### PR TITLE
Reimplement Popover with floating-ui

### PR DESCRIPTION
**Pull request by sergiou87**
Based on #16640

## Description

This PR reimplements the `Popover` component to use [floating-ui](https://floating-ui.com/). The idea is to have a single reusable implementation for floating elements (like the balloon popovers or the autocomplete list) that is also responsive in different situations, like when the UI is scaled and content needs to adapt to remain usable and accessible.

This PR also adds the use of this new `Popover` to display the autocomplete list of `AutocompletingTextInput`.

The changes for every existing popover look like this:
- Remove explicit positioning from CSS
- If necessary, wrap the anchor element so we can extract a ref of it
- Use its ref as anchor

Some other changes made to the popover:
- Some of its layout properties like `position` or `zIndex` have been moved to TypeScript
- The tip is now made of two `div`s instead of using the `::before` and `::after` pseudo-elements
- Instead of positioning the tip, now we position the whole popover in relation to the anchor element, which works like floating-ui's API.
- The tip is not positioned in a corner anymore. Instead, with the new positioning approach described above, we will decide to which edges of the anchor element we'll align the popover, and the tip will show up centered on the anchor.
- I've renamed caret to tip because we're using caret for the text cursor and it was driving me crazy 😂 

Regarding the autocomplete popup, I created an invisible `div` element that represents the text caret, so that we can anchor the popover to it.

⚠️ The branch dropdown popover in the "Preview Pull Request" dialog still has some logic to calculate its height, something this new popover should take care of (to some extent, at least), mainly because it involves virtualized lists and is complex enough to deserve its own PR.

### Screenshots

https://github.com/desktop/desktop/assets/1083228/e3366057-fe08-4076-bdeb-cde35c0d5fac

## Release notes

Notes: [Improved] Make floating elements more responsive as the window or the UI are resized

### CODE REVIEW

File: app/styles/ui/_commit-message-avatar.scss
Comment (sergiou87) [inline app/styles/ui/_commit-message-avatar.scss:1]: The changes in this file just intend to get rid of some unnecessary margins that were messing with the location of the popover:
| Before | After |
|--------|--------|
| image | image |

File: app/styles/ui/changes/_commit-message.scss
Comment (sergiou87) [inline app/styles/ui/changes/_commit-message.scss:1]: This is also part of fixing the layout of the avatar in the commit message area.

File: app/styles/ui/history/_commit-list.scss [hunk 1]
```diff
@@ -6,27 +6,6 @@
...
-      color: var(--call-to-action-bubble-color);
-      padding: 1px 5px;
-      border-radius: var(--border-radius);
-      margin-left: var(--spacing-third);
-    }
-    img {
-      width: 100%;
-    }
-  }
-
```
Comment (sergiou87) [inline app/styles/ui/history/_commit-list.scss:29]: These styles belong to the intro popovers introduced in #11737 and removed in #14463

**Comment by sergiou87** [comment]
As I was working on using this for the autocomplete popup, I ran into a few issues that should be fixed here. Nothing huge but still…

**Review by tidy-dev — APPROVED** [review]
🏆 This was a ton of effort and code changes.. the diff area ones looked like a particular pain! (🙈 I didn't realize the trouble/can of worms I was opening when I suggested the floating-ui) Thank you for getting us these awesome popovers. Playing with the pointer moving with a scroller container on the git avatars was so smooth!

I tested the various popovers out. They all worked great with one exception and it is a zoom clipping issue on whitespace popover. I would have thought float-ui would have automatically taken care of it, but maybe not since it is the diff? Any which ways, could definitely be looked into in another PR as when not zoomed in, it is great.

Zoom: 175% clipping.
CleanShot 2023-05-22 at 15 17 21

**Review by tidy-dev — APPROVED** [review]
🏆 This was a ton of effort and code changes.. the diff area ones looked like a particular pain! (🙈 I didn't realize the trouble/can of worms I was opening when I suggested the floating-ui) Thank you for getting us these awesome popovers. Playing with the pointer moving with a scroller container on the git avatars was so smooth!

I tested the various popovers out. They all worked great with one exception and it is a zoom clipping issue on whitespace popover. I would have thought float-ui would have automatically taken care of it, but maybe not since it is the diff? Any which ways, could definitely be looked into in another PR as when not zoomed in, it is great.

Zoom: 175% clipping.
CleanShot 2023-05-22 at 15 17 21

**Pull request merged**